### PR TITLE
WIP #57: first pass at a minimal emacs config for lsp-mode typescript

### PR DESCRIPTION
**Pull request by e-carlin**
This is a WIP config for using lsp-mode for typescript. This is about as bare-bones as you can make lsp-mode. In my personal config I have more things turned on (ex descriptions for coding errors, auto-completion menus, etc). I'm by no means an lsp-mode expert so this almost certainly needs refinement.

One area that would be nice to improve on is setup. Once you check out this code and open a .tsx file you'll get prompted to answer some questions. It would be nice to set the answers for the questions in the config. 

I would upload screenshots of the questions but GH won't accept image uploads for me right now. So I'll describe them in text:

1. `<filename> is not part of any project`: Enter `I Import project by selecting root interactively`. Enter ~/src/radiasoft/sirepo/react/src
2. `Unable to find installed server supporting this file. The following servers could be installed automatically`: Enter ts-ls. 

You should now have syntax highlighting and indentation. 

There are other lsp features that I like. The main one is `M-x lsp-find-definition`. If you move your cursor over a piece of code (ex an imported name) and then do `M-x lsp-find-definition` it will jump to the definition of that name. 

We can enable it for other languages (ex python) but just starting out with typescript. Lmk what you think.

### CODE REVIEW

**Comment by e-carlin** [comment]
FWIW - If you answer one of the questions "wrong" you can `rm -rf ~/.emacs.d/.cache/lsp`, close emacs, re-open, and you will be prompted again.

File: emacs.d/b-lsp.el [hunk 1]
```diff
@@ -0,0 +1,19 @@
+(provide 'b-lsp)
+
+(require 'package)
+(setq package-user-dir (concat (file-name-directory load-file-name) "melpa"))
```
Comment (robnagler) [inline emacs.d/b-lsp.el:4]: Should this be in bivio-init.el?

File: emacs.d/b-lsp.el [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
+(package-initialize)
+
+(require 'lsp-mode)
+(require 'lsp-pyright)
+(setq lsp-completion-provider :none)
+(setq lsp-pyright-auto-import-completions
+      nil)
+(setq lsp-signature-auto-activate nil) ;; disalbe bottom documentation popup
```
Comment (robnagler) [inline emacs.d/b-lsp.el:13]: typo and don't side comment.

File: emacs.d/b-lsp.el [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+(setq company-dabbrev-downcase 1)
+(setq company-idle-delay 0.01)
+(add-hook 'js-mode-hook #'lsp)
+(add-hook 'web-mode-hook #'lsp)
+(setq gc-cons-threshold 100000000)
```
Comment (robnagler) [inline emacs.d/b-lsp.el:18]: definitely in bivio-init.el

File: emacs.d/b-lsp.el [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+(add-to-list 'package-archives '("melpa" . "https://melpa.org/packages/") t)
+(package-initialize)
+
+(require 'lsp-mode)
+(require 'lsp-pyright)
+(setq lsp-completion-provider :none)
+(setq lsp-pyright-auto-import-completions
+      nil)
+(setq lsp-signature-auto-activate nil) ;; disalbe bottom documentation popup
+(setq company-dabbrev-downcase 1)
```
Comment (robnagler) [inline emacs.d/b-lsp.el:14]: I had to look this up. Really odd name. Is company mode something we should use instead of dabbrev?

**Review by robnagler — CHANGES_REQUESTED** [review]

File: emacs.d/b-lsp.el (hunk 1 above)
Comment (e-carlin) [reply emacs.d/b-lsp.el:4]: Up to you. My intention with putting everything in `b-lsp.el` was that if you decide to remove lsp in the future you can just delete this file. If you spread out the configuration then you'll have to find each bit and remove it. In my own config I've been slowly ripping things out of my main init file and into files like this. But, its just personal preference.

File: emacs.d/b-lsp.el (hunk 4 above)
Comment (e-carlin) [reply emacs.d/b-lsp.el:14]: Sorry, those were copy past bugs from my own config. 7a1ec86

I like the company-mode completions. It looks more "IDE like" in that you get a dropdown menu of completion and can cycle through them. But, not mandatory for lsp-mode.

File: emacs.d/b-lsp.el (hunk 3 above)
Comment (e-carlin) [reply emacs.d/b-lsp.el:18]: Should `read-process-output-max` go over too? How do you decide what you put in bivio-init.el vs here?

File: emacs.d/b-lsp.el (hunk 2 above)
Comment (e-carlin) [reply emacs.d/b-lsp.el]: 8715b8e9423034825ab8fe6e8a99d3f9b79c9561

File: emacs.d/b-lsp.el (hunk 1 above)
Comment (robnagler) [reply emacs.d/b-lsp.el:4]: Makes sense leave as is.

File: emacs.d/b-lsp.el (hunk 3 above)
Comment (robnagler) [reply emacs.d/b-lsp.el:17]: ditto.

File: emacs.d/b-lsp.el (hunk 4 above)
Comment (robnagler) [reply emacs.d/b-lsp.el]: Thanks. I have seen that. It could be useful. Perhaps when there's time we'll look into it.

**Review by robnagler — APPROVED** [review]
I'll let you merge

**Comment by e-carlin** [comment]
@moellep I don't want to merge until you've tested. If you checkout this branch and restart emacs then the next time you open a tsx file you'll be in `lsp-mode`.

**Comment by moellep** [comment]
I tried it now and it indents correctly which is great. When I open a tsx file, it does prompt me with this message, which I need to dismiss initially:

```Unable to find installed server supporting this file. The following servers could be installed automatically:```
and then I hit enter and this error is shown:
```File mode specification error: (wront-type-argument lsp--client nil)```

**Comment by e-carlin** [comment]
@moellep that message is to install a language server. Ideally we could set in our configuration what language server we want installed but it doesn't seem like there is support for that. 

I've created a movie showing how to set things up. You may need to run `rm -rf ~/.emacs.d/.cache/lsp/` to get the prompts to show up again.

The steps are:
1. Select a language server to be installed. You can hit tab to see the options. I find `ts-ls` works the best so that is what I enter.
2. Select a  "workspace root". The language server will "watch" all files from this directory down. I choose `I -> Import project by setting root interactively`. I then enter `~/src/radiasoft/sirepo/react/src` which contains all of our tsx files.

This setup is just one time. The first time you open a tsx file it will ask for these details. If you ever open a tsx file outside of the workspace root (`~/src/radiasoft/sirepo/react/src`) lsp-mode will ask you if you want to add a new workspace to be watched. You can watch as many workspaces as you want.

Is this manual configuration ok or should I do some more digging and try to find a way to get it setup automatically?

https://user-images.githubusercontent.com/10264515/220968283-dde5153c-4e24-4aee-8604-3d4f3841849a.mp4

**Comment by moellep** [comment]
Thanks Evan - I followed the steps and it doesn't prompt me on each file anymore.

**Pull request merged**